File: js/complete/no021.js
#!/usr/bin/env node

var fns = require('../functions.js'),
    operator = require('../operator.js'),
    timer = require('../timer.js');

exports.main = function(verbose) {
    if (typeof verbose == 'undefined') {
        verbose = false;
    }
    var n = 10000,
        factors = fns.allFactors(n - 1);

    // sum of proper divisors
    var firstPass = [], i;
    for (i = 1; i < n; i++) {
        firstPass.push(operator.sum(factors[i]) - 1);
    }
    var maxOut = Math.max.apply(Math, firstPass);

    factors = fns.allFactors(maxOut, factors);
    /* applying sum of divisors twice to i we have
       s_1 = func(i), s_2 = func(s_1)
       s_1 = sum(factors[i]) - i, s_2 = sum(factors[s_1]) - s_1
       i == s_2 <==>
       i == sum(factors[sum(factors[i]) - i]) - (sum(factors[i]) - i) <==>
       sum(factors[sum(factors[i]) - i]) == sum(factors[i])
       Similarly s_1 != i <==> sum(factors[i]) != 2 * i */
    var result = [], factorSum;
    for (i = 2; i < n; i++) {
        factorSum = operator.sum(factors[i]);
        if (factorSum != 2 * i) {
            if (operator.sum(factors[factorSum - i]) == factorSum) {
                result.push(i);
            }
        }
    }

    if (verbose) {
        return [operator.sum(result), '.\nThe full list of such amicable numbers is ',
                result.join(', '), '.'].join('');
    } else {
        return operator.sum(result);
    }
};

if (require.main === module) {
    timer.timer(21, exports.main);
}

File: python/complete/no021.py
#!/usr/bin/env python

from python.decorators import euler_timer
from python.functions import all_factors


def main(verbose=False):
    n = 10000
    factors = all_factors(n - 1)

    # sum of proper divisors
    first_pass = [sum(factors[i]) - i for i in range(1, n)]
    max_out = max(first_pass)

    factors = all_factors(max_out, factors)

    # applying sum of divisors twice to i we have
    # s_1 = func(i), s_2 = func(s_1)
    # s_1 = sum(factors[i]) - i, s_2 = sum(factors[s_1]) - s_1
    # i == s_2 <==>
    # i == sum(factors[sum(factors[i]) - i]) - (sum(factors[i]) - i) <==>
    # sum(factors[sum(factors[i]) - i]) == sum(factors[i])
    # Similarly s_1 != i <==> sum(factors[i]) != 2*i
    result = [i for i in range(2, n) if
              sum(factors[sum(factors[i]) - i]) == sum(factors[i]) and
              sum(factors[i]) != 2 * i]

    if verbose:
        return '%s.\nThe full list of such amicable numbers is %s.' % (
            sum(result), ', '.join(str(elt) for elt in result))
    else:
        return sum(result)

if __name__ == '__main__':
    print euler_timer(21)(main)(verbose=True)
